### PR TITLE
#76 Provides logging for InMemoryCacheActionFactory

### LINKED ISSUE #76: InMemoryCacheAction log
**Is your feature request related to a problem? Please describe.**
Add an [action node log](https://github.com/Knotx/knotx-fragments/pull/55) support for In-memory Cache Action.

**Describe the solution you'd like**
The log should contain:
- cache key (`key`)
- cache value if available (`cacheValue`)
- value from `doAction` if not found in the cache (`newValue`)

All the above entries should appear when `INFO` log level is set.

**Additional context**
Consider the situation when `doAction` action throws some error or responds with `_error` transition. Check also [this PR](https://github.com/Knotx/knotx-fragments/pull/45/files#diff-d48decbc78ea648abf1a06fcb9681288).

### PR DESCRIPTION
**Pull request by marcinus**
Provides logging for InMemoryCacheActionFactory

## Description
The logs are applied on these events:
-> Whenever there is a cache lookup
-> On cache miss (no value in cache, successfully computed)
-> On cache hit (value present in cache)
-> On cache pass (meaning there was no payload to be cached - either failure or just no data)

## Motivation and Context
Fixes #76 
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the [code style](/CONTRIBUTING.md#coding-conventions) of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

---
I hereby agree to the terms of the Knot.x Contributor License Agreement.

### CODE REVIEW

File: handler/core/src/main/java/io/knotx/fragments/handler/action/InMemoryCacheActionFactory.java [hunk 1]
```diff
@@ -72,44 +84,85 @@ public Action create(String alias, JsonObject config, Vertx vertx, Action doActi
...
-
+        ActionLogger actionLogger = ActionLogger.create(alias, logLevel);
         String cacheKey = getCacheKey(config, fragmentContext.getClientRequest());
-        Object cachedValue = cache.getIfPresent(cacheKey);
+
+        getFromCache(fragmentContext, cacheKey, actionLogger)
+            .switchIfEmpty(callDoActionAndCache(fragmentContext, cacheKey, actionLogger))
+            .map(Future::succeededFuture)
+            .onErrorReturn(Future::failedFuture)
+            .doOnSuccess(future -> future.setHandler(resultHandler))
```
Comment (tomaszmichalak) [inline handler/core/src/main/java/io/knotx/fragments/handler/action/InMemoryCacheActionFactory.java]: Nice :)

File: handler/core/src/main/java/io/knotx/fragments/handler/action/InMemoryCacheActionFactory.java [hunk 2]
```diff
@@ -147,4 +200,31 @@ private SourceDefinitions buildSourceDefinitions(ClientRequest clientRequest) {
...
+  private static void logCacheHit(ActionLogger actionLogger, Object cachedValue) {
+    actionLogger.info(CACHE_HIT, cachedValue);
+  }
+
+  private static void logCacheMiss(ActionLogger actionLogger, Object computedValue) {
+    actionLogger.info(CACHE_MISS, computedValue);
+  }
+
+  private static void logCachePass(ActionLogger actionLogger, FragmentResult failedResult) {
+    actionLogger.info(CACHE_PASS,
```
Comment (tomaszmichalak) [inline handler/core/src/main/java/io/knotx/fragments/handler/action/InMemoryCacheActionFactory.java]: Can we log also node log here? + unit test

File: handler/core/src/test/java/io/knotx/fragments/handler/action/logging/InMemoryCacheActionFactoryLoggingTest.java [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+import io.vertx.core.json.JsonObject;
+import io.vertx.junit5.VertxTestContext;
+import java.util.concurrent.TimeUnit;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+@ExtendWith(KnotxExtension.class)
+class InMemoryCacheActionFactoryLoggingTest {
```
Comment (tomaszmichalak) [inline handler/core/src/test/java/io/knotx/fragments/handler/action/logging/InMemoryCacheActionFactoryLoggingTest.java]: All cases focus on the INFO log level. Can we check that there are no node log entries when the action (in-memory action) ends with _success transition and log level is ERROR.

File: handler/core/src/main/java/io/knotx/fragments/handler/action/InMemoryCacheActionFactory.java [hunk 4]
```diff
@@ -50,13 +53,22 @@
...
  *       payloadKey = product
+ *       logLevel = info
  *     }
  *   }
  * </pre>
  */
 @Cacheable
 public class InMemoryCacheActionFactory implements ActionFactory {
 
+  public static final String CACHE_LOOKUP = "CACHE_LOOKUP";
```
Comment (tomaszmichalak) [inline handler/core/src/main/java/io/knotx/fragments/handler/action/InMemoryCacheActionFactory.java]: Should be public?

File: handler/core/src/main/java/io/knotx/fragments/handler/action/InMemoryCacheActionFactory.java (hunk 2 above)
Comment (marcinus) [reply handler/core/src/main/java/io/knotx/fragments/handler/action/InMemoryCacheActionFactory.java]: Node log will be provided by a call to `logDoAction` in `callDoActionAndCache`

File: handler/core/src/main/java/io/knotx/fragments/handler/action/InMemoryCacheActionFactory.java [hunk 5]
```diff
@@ -50,13 +53,22 @@
  *       }
  *       cacheKey = product-{param.id}
  *       payloadKey = product
+ *       logLevel = info
  *     }
  *   }
  * </pre>
  */
 @Cacheable
 public class InMemoryCacheActionFactory implements ActionFactory {
```
Comment (tomaszmichalak) [inline handler/core/src/main/java/io/knotx/fragments/handler/action/InMemoryCacheActionFactory.java]: Please update README & changelog.

File: handler/core/src/test/java/io/knotx/fragments/handler/action/logging/InMemoryCacheActionFactoryLoggingTest.java [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+
+    assertTrue(testContext.awaitCompletion(5, TimeUnit.SECONDS));
+    if (testContext.failed()) {
+      throw testContext.causeOfFailure();
+    }
+  }
+
+  @DisplayName("Cache operations are not logged when log level is error")
+  @Test
+  void callActionWithSuccessTransitionWithLogLevelError(VertxTestContext testContext)
```
Comment (tomaszmichalak) [inline handler/core/src/test/java/io/knotx/fragments/handler/action/logging/InMemoryCacheActionFactoryLoggingTest.java]: This test is duplicated, can we implement parameterized test? (see https://junit.org/junit5/docs/current/user-guide/#writing-tests-parameterized-tests)

Examples: https://github.com/Knotx/knotx-fragments/blob/master/handler/core/src/test/java/io/knotx/fragments/handler/action/cb/CircuitBreakerActionFactoryTest.java#L95

File: handler/core/src/test/java/io/knotx/fragments/handler/action/logging/InMemoryCacheActionFactoryLoggingTest.java [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+          });
+        });
+
+    assertTrue(testContext.awaitCompletion(5, TimeUnit.SECONDS));
+    if (testContext.failed()) {
+      throw testContext.causeOfFailure();
+    }
+  }
+
+  @DisplayName("Cache pass gets logged when no payload is added by doAction when log level is info")
```
Comment (tomaszmichalak) [inline handler/core/src/test/java/io/knotx/fragments/handler/action/logging/InMemoryCacheActionFactoryLoggingTest.java]: Should we log `cache pass` when log level is ERROR?

File: handler/core/src/test/java/io/knotx/fragments/handler/action/logging/InMemoryCacheActionFactoryLoggingTest.java (hunk 6 above)
Comment (marcinus) [reply handler/core/src/test/java/io/knotx/fragments/handler/action/logging/InMemoryCacheActionFactoryLoggingTest.java]: Good idea, will update the PR

**Pull request merged**